linkerd/detect/src/lib.rs

```diff
@@ -27,28 +27,30 @@ pub trait Detect<I>: Clone + Send + Sync + 'static {
 pub type DetectResult<P> = Result<Option<P>, DetectTimeoutError<P>>;
 
 #[derive(Error)]
-#[error("{} protocol detection timed out after {:?}", std::any::type_name::<P>(), .0)]
+#[error("{} protocol detection timed out after {0:?}", std::any::type_name::<P>())]
 pub struct DetectTimeoutError<P>(time::Duration, std::marker::PhantomData<P>);
 
 #[derive(Copy, Clone, Debug)]
-pub struct NewDetectService<D, N> {
+pub struct Config<D> {
+    pub detect: D,
+    pub capacity: usize,
+    pub timeout: time::Duration,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct NewDetectService<P, D, N> {
     inner: N,
-    detect: D,
-    capacity: usize,
-    timeout: time::Duration,
+    params: P,
+    _detect: std::marker::PhantomData<fn() -> D>,
```

Comment:
This is the only small question that I have: are we using `PhantomData` here to satisfy the compiler? Is this a way of saying that the service will live at least as long as the config itself, or am I misreading it?

I looked through the [docs](https://doc.rust-lang.org/std/marker/struct.PhantomData.html) to see what `PhantomData` does, and also searched the webs. Most people seem to point that it's used to express lifetime dependencies so just wanted to see if I get this right.

Reply:
> This is the only small question that I have: are we using PhantomData here to satisfy the compiler? Is this a way of saying that the service will live at least as long as the config itself, or am I misreading it?

It's not really about lifetime in the ownership/borrowing sense and more about our ability to know the `D` type. Specifically here:

https://github.com/linkerd/linkerd2-proxy/blob/fa5c53becfa93995d7107ada229421bc37bc6bb0/linkerd/detect/src/lib.rs#L97-L111

The `NewService` implementation needs to know about the `D` type (to build the service type) -- but if we tried to be generic over `D` without it being a generic type on the _new_ type, then we'd get a compile error saying that `D` is unconstrained. `PhantomData` lets us bind a `D` type on `NewDetect` so that we can use it for the `NewService` implementation. It also means that we have to _know_ what the `D` type is when we build this layer and if it's ambiguous, we may need to document it explicitly.

It's generally best to avoid marker types like this if at all possible, but they're a necessary evil in situations like this.

Also, we use a marker type like `PhatnomData<fn() -> D>` (instead of just `PhantomData<D>` because this impacts the `Send`/etc requirements on `D` -- `NewDetect` can be `Send` (etc) without `D` having to be documented as such. I think about it as "we are able to produce a `D`" -- because that's what `extract_param` does here.